src/server/src/main/java/io/cassandrareaper/storage/memory/ReplicaLockManagerWithTtl.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2024-2024 DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cassandrareaper.storage.memory;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class ReplicaLockManagerWithTtl {
+
+  private final ConcurrentHashMap<String, LockInfo> replicaLocks = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<UUID, Set<UUID>> runToSegmentLocks = new ConcurrentHashMap<>();
+  private final Lock lock = new ReentrantLock();
```

Comment:
I hear you on the premature optimisation angle so I'm happy to run with what you have here if you feel it's best. I do note that the cleanups are happening every second, which is where most of my concern comes from.

Reply:
> Between the moment when they check if replicas are locked and the moment they lock these replicas, there could be a race condition.

Roger, I've been mulling this over and I can see what you're saying, although I still think you could make better uses of the concurrent map to avoid locking so early using an external lock. 

I think I'm getting a bit confused between the `Lock` in the Java sense in one place and talking about our own `Locks` which exist in a table. In the first case, it is an atomic CAS operation but in the latter it is a read then a write from the Java side. I can see that in that latter case you might end up with an overwrite using your current logic. 

That said, I think a revision could be made to use a CAS type operation just using the features of the `ConcurrentHashMap`. For example, `putIfAbsent` would probably enable an atomic read then write operation without recourse to the external lock.

Reply:
As discussed we won't have much contention anyway, everything happens in memory and we have a safe implementation for setting our distributed lock the way it's implemented here.
It'll work fine even with many competing threads (which we won't have).

Reply:
As discussed elsewhere, the fact that a segment can run on multiple replicas probably we need a more complex data structure than either a one way map or nested maps. We would investigate this later if required.